### PR TITLE
Add Transport property to Request for metric tags

### PR DESCRIPTION
**Pull request by zmt**
This change allows us to add a tag to each of the currently emitted metrics so we can track requests by transport. The tchannel, http, and grpc inbound transports set the transport property. If the transport property is not set, the transport tag will be marked "unknown". This will be true for all callers and for any third-party inbound transports that do not set the transport property.

### CODE REVIEW

File: internal/observability/middleware_test.go [hunk 1]
```diff
@@ -60,6 +61,7 @@ func TestMiddlewareLogging(t *testing.T) {
 		return []zapcore.Field{
 			zap.String("source", req.Caller),
 			zap.String("dest", req.Service),
+			zap.String("transport", string(req.Transport)),
```
Comment (kriskowal) [inline internal/observability/middleware_test.go]: String cast is superfluous here.

File: internal/observability/middleware_test.go [hunk 2]
```diff
@@ -304,9 +307,11 @@ func TestMiddlewareMetrics(t *testing.T) {
 	}
 
 	for _, tt := range tests {
+		t.Log(tt)
```
Comment (kriskowal) [inline internal/observability/middleware_test.go]: Debug junk

File: internal/observability/middleware_test.go [hunk 3]
```diff
@@ -304,9 +307,11 @@ func TestMiddlewareMetrics(t *testing.T) {
...
 		validate := func(mw *Middleware, direction string) {
 			key, free := getKey(req, direction)
 			edge := mw.graph.getEdge(key)
+			t.Log(edge)
```
Comment (kriskowal) [inline internal/observability/middleware_test.go]: Also debug junk

File: internal/observability/middleware_test.go [hunk 4]
```diff
@@ -429,6 +438,7 @@ func TestMiddlewareSuccessSnapshot(t *testing.T) {
 	tags := metrics.Tags{
 		"dest":             "service",
 		"direction":        "inbound",
+		"transport":        "default",
```
Comment (kriskowal) [inline internal/observability/middleware_test.go]: Let’s trace this and change it to "unknown", as you suggest.

File: internal/observability/middleware_test.go (hunk 1 above)
Comment (zmt) [reply internal/observability/middleware_test.go]: bad copypasta

**Comment by zmt** [comment]
> Let’s ask @akshayjshah whether it’s safe to add a tag to an existing metric (same name).

Done. Answer is that it is the responsibility of the consumer to adjust for additional tags. This should be expected behavior.

File: internal/observability/middleware_test.go (hunk 4 above)
Comment (zmt) [reply internal/observability/middleware_test.go]: https://sourcegraph.com/github.com/yarpc/metrics@1e19de5b971489a45d178e12a0f72a78c70e300e/-/blob/tag.go#L34:2$references

File: internal/observability/middleware_test.go [hunk 5]
```diff
@@ -532,3 +543,39 @@ func TestMiddlewareFailureSnapshot(t *testing.T) {
...
+
+func TestFixDefaultTransportValue(t *testing.T) {
+	tests := []struct {
+		desc      string
+		transport string
+		expected  string
+	}{
+		{
+			desc:      "unknown",
+			transport: "",
```
Comment (AlexAPB) [inline internal/observability/middleware_test.go]: nit: drop this line, you can rely on go defaults; "" for strings

File: internal/observability/graph.go [hunk 6]
```diff
@@ -239,3 +242,11 @@ func newEdge(logger *zap.Logger, meter *metrics.Scope, req *transport.Request, d
 		serverErrLatencies: serverErrLatencies,
 	}
 }
+
+// fixDefaultTransportValue works around hard-coded default value of "default" in go.uber.org/net/metrics
+func fixDefaultTransportValue(t string) string {
```
[the remark this answers is not in the document]
Comment (zmt) [reply internal/observability/graph.go]: Done.

File: internal/observability/middleware_test.go (hunk 5 above)
Comment (zmt) [reply internal/observability/middleware_test.go]: Done.

File: internal/observability/graph_test.go [hunk 7]
```diff
@@ -38,6 +38,7 @@ func TestEdgeNopFallbacks(t *testing.T) {
 	req := &transport.Request{
 		Caller:          "caller",
 		Service:         "service",
+		Transport:       unknownIfEmpty(""),
```
Comment (akshayjshah) [inline internal/observability/graph_test.go]: This is just "unknown" - why hop through the extra function call?

File: internal/observability/middleware_test.go [hunk 8]
```diff
@@ -429,6 +436,7 @@ func TestMiddlewareSuccessSnapshot(t *testing.T) {
 	tags := metrics.Tags{
 		"dest":             "service",
 		"direction":        "inbound",
+		"transport":        unknownIfEmpty(""),
```
Comment (akshayjshah) [inline internal/observability/middleware_test.go]: Likewise - why the extra function call?

File: internal/observability/middleware_test.go [hunk 9]
```diff
@@ -532,3 +543,38 @@ func TestMiddlewareFailureSnapshot(t *testing.T) {
...
+		},
+		{
+			desc:      "http",
+			transport: "http",
+			expected:  "http",
+		},
+		{
+			desc:      "grpc",
+			transport: "grpc",
+			expected:  "grpc",
```
Comment (akshayjshah) [inline internal/observability/middleware_test.go]: `desc` and `expected` are always the same - why not name all the subtests using `tt.expected`?

File: transport/tchannel/handler.go [hunk 10]
```diff
@@ -132,6 +132,7 @@ func (h handler) callHandler(ctx context.Context, call inboundCall, responseWrit
 		Caller:          call.CallerName(),
 		Service:         call.ServiceName(),
 		Encoding:        transport.Encoding(call.Format()),
+		Transport:       "tchannel",
```
Comment (akshayjshah) [inline transport/tchannel/handler.go]: Can we make these constants?

Comment (zmt) [reply transport/tchannel/handler.go]: These strings already existed as transportName in each transport package, so I used those.

File: internal/observability/middleware_test.go (hunk 9 above)
Comment (zmt) [reply internal/observability/middleware_test.go]: done and moved to graph_test.go where it belongs

File: internal/observability/middleware_test.go [hunk 11]
```diff
@@ -532,3 +543,38 @@ func TestMiddlewareFailureSnapshot(t *testing.T) {
 	}
 	assert.Equal(t, want, snap, "Unexpected snapshot of metrics.")
 }
+
+func TestFixDefaultTransportValue(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (zmt) [reply internal/observability/middleware_test.go]: done

File: internal/observability/middleware_test.go (hunk 8 above)
Comment (zmt) [reply internal/observability/middleware_test.go]: done

File: internal/observability/graph_test.go (hunk 7 above)
Comment (zmt) [reply internal/observability/graph_test.go]: done

File: internal/observability/middleware_test.go [hunk 12]
```diff
@@ -60,6 +61,7 @@ func TestMiddlewareLogging(t *testing.T) {
 		return []zapcore.Field{
 			zap.String("source", req.Caller),
 			zap.String("dest", req.Service),
+			zap.String("transport", unknownIfEmpty(req.Transport)),
```
Comment (akshayjshah) [inline internal/observability/middleware_test.go]: For consistency, all assertions should hard-code a string or all should use `unknownIfEmpty`. We're doing one here and a different thing below.

File: transport/grpc/constants.go [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package grpc
+
+const transportName = "grpc"
```
Comment (akshayjshah) [inline transport/grpc/constants.go]: nit: we're not religious about this in YARPC, but our style guide recommends underscore-prefixing globals.

File: internal/observability/middleware_test.go (hunk 12 above)
Comment (akshayjshah) [reply internal/observability/middleware_test.go]: Never mind, I'm wrong.

**Pull request merged**